Add Docker image

fracek opened the pull request:
I created a docker image so it's easier to run the node. It can also persist data between restarts.

Mirko-von-Leipzig commented:
Hmm okay so our CI secrets don't like external contributors. Will have to figure that out.

fracek commented:
Ok I rebased it.

fracek commented:
As mentioned on Discord there seems to be some issues with the image. The node is synchronizing but the RPC module is not working.

CHr15F0x reviewed (commented):
As I mentioned on Discord, I tried running the image and it turns out sync works but I cannot see rpc starting. I haven't investigated yet.

fracek commented:
I updated the example and now I can execute `rpc_examples.sh` using the docker image. The only outstanding issue is that the container cannot be gracefully exited with ctrl-c, but needs to be killed with `docker kill`.

aphelionz commented on Dockerfile:
```diff
@@ -0,0 +1,42 @@
+
+# Create directory and volume for persistent data
+RUN mkdir -p /usr/share/pathfinder/data
+
```
We shouldn't run this container as root. It would be good to toss a `USER 1000` in here and ensure all the right things still work.

fracek replied:
You're right, I'll fix this.

aphelionz replied:
Thank you!

CHr15F0x replied:
@fracek I've tested your fixes and everything works for me :tada: . I'l retest when the latest touches are in (ie. running as non-root).

aphelionz commented on Dockerfile:
```diff
@@ -0,0 +1,44 @@
+# python environment and scripts
+# This docker image is not using virtual environments, so replicate
+# the directory structure with a shim python script.
+ENV VIRTUAL_ENV=/usr/share/pathfinder/py/venv
```
@Mirko-von-Leipzig @CHr15F0x Is there a way we can change the Rust code so that it reads from the environments `python3` by default i.e. `#!/usr/bin/env python3`, or from configuration?

Mirko-von-Leipzig replied:
I don't really understand these `venv` considerations or restrictions so I'm not really the best person to ask 😅

CHr15F0x replied:
I'll investigate, np :+1: .

aphelionz commented:
@fracek can we try this? I spent some time with alpine and I think its worthwhile. feel free to use this in your PR and still get credit 💪 

```
########################################
# Stage 1: Build the pathfinder binary #
########################################
FROM rust:1.59-alpine AS rust-builder

RUN apk add --no-cache musl-dev gcc openssl-dev

WORKDIR /usr/src/pathfinder

# Build only the dependencies first. This utilizes
# container layer caching for Rust builds
RUN mkdir crates
RUN cargo new --lib crates/pedersen
# Correct: --lib. We'll handle the binary later.
RUN cargo new --lib crates/pathfinder
COPY Cargo.toml Cargo.toml

COPY crates/pathfinder/Cargo.toml crates/pathfinder/Cargo.toml
COPY crates/pedersen/Cargo.toml crates/pedersen/Cargo.toml
COPY crates/pedersen/benches crates/pedersen/benches

RUN RUSTFLAGS='-L/usr/lib -Ctarget-feature=-crt-static' cargo build --target=x86_64-unknown-linux-musl


# Compile the actual libraries and binary now
COPY . .

# Mark these for re-compilation 
RUN touch crates/pathfinder/src/lib.rs
RUN touch crates/pedersen/src/lib.rs

RUN RUSTFLAGS='-L/usr/lib -Ctarget-feature=-crt-static' cargo build --target=x86_64-unknown-linux-musl --bin pathfinder

#######################################
# Stage 2: Build the Python libraries #
#######################################
FROM python:3.8-alpine AS python-builder

RUN apk add --no-cache gcc musl-dev gmp-dev g++

WORKDIR /usr/share/pathfinder
COPY py py
RUN python3 -m pip install --upgrade pip
RUN python3 -m pip install -r py/requirements-dev.txt

# This reduces the size of the python libs by about 50%
ENV PY_PATH=/usr/local/lib/python3.8/
RUN find ${PY_PATH} -type d -a -name test -exec rm -rf '{}' +
RUN find ${PY_PATH} -type d -a -name tests  -exec rm -rf '{}' +
RUN find ${PY_PATH} -type f -a -name '*.pyc' -exec rm -rf '{}' +
RUN find ${PY_PATH} -type f -a -name '*.pyo' -exec rm -rf '{}' +

# Pathfinder expects a very specific directory structure for the
# python environment and scripts
# This docker image is not using virtual environments, so replicate
# the directory structure with a shim python script.
ENV VIRTUAL_ENV=/usr/share/pathfinder/py/venv
RUN mkdir -p ${VIRTUAL_ENV}/bin
RUN ln -s /usr/local/bin/python3 ${VIRTUAL_ENV}/bin/python 
RUN ${VIRTUAL_ENV}/bin/python --version

#######################
# Final Stage: Runner #
#######################
FROM python:3.8-alpine AS runner

RUN apk add --no-cache tini

COPY --from=rust-builder ["/usr/lib/libstdc++.so.6", "/usr/lib/libgcc_s.so.1", "/usr/lib/" ]
COPY --from=rust-builder /usr/src/pathfinder/target/x86_64-unknown-linux-musl/debug/pathfinder /usr/local/bin/pathfinder
COPY --from=python-builder /usr/local/lib/python3.8/ /usr/local/lib/python3.8/
COPY --from=python-builder /usr/share/pathfinder/py/venv /usr/share/pathfinder/py/venv

ENV VIRTUAL_ENV=/usr/share/pathfinder/py/venv
COPY py/src /usr/share/pathfinder/py/src

# Create directory and volume for persistent data
RUN mkdir -p /usr/share/pathfinder/data
RUN chown 1000:1000 /usr/share/pathfinder/data
VOLUME /usr/share/pathfinder/data

USER 1000:1000
EXPOSE 9545
WORKDIR /usr/share/pathfinder/data

ENTRYPOINT ["tini", "--"]
CMD ["/usr/local/bin/pathfinder"]
```

This:
1. fixes the ctrl+c problem by using `tini`
2. Keeps `USER 1000`
3. Reduces the image size to 380MB from 688MB (and I'm sure we can find more optimizations too)

aphelionz commented on README.md:
````diff
@@ -139,6 +139,27 @@ RUST_LOG=pathfinder=<log level> cargo run --release <path-to-pathfinder-repo> --
+You can then start the node with:
+
+```bash
+docker run --rm -it -e RUST_LOG=info -p 9545:9545 pathfinder --http-rpc "0.0.0.0:9545" --ethereum.url  https://goerli.infura.io/v3/<project-id>
````
These commands will need `--ethereum.password` as well, just placeholders. Thank you!

aphelionz commented:
@fracek heads up #188 was just merged which eliminates the need for the venv hack in the `Dockerfile` 

At this point it might make sense for you to make a fresh PR with the alpine version sans the `venv` lines. What do you think?

fracek commented:
@aphelionz I can rebase and amend this PR.

aphelionz commented:
Sounds good to me, thank you!

fracek commented:
I updated the Dockerfile. I changed three things from what you posted:

* Removed references to the virtual environment
* Copy libgmp to the runner image
* Add pathfinder to `ENTRYPOINT` so users can pass the arguments directly to docker run

aphelionz reviewed (approved):
LGTM! Thanks @fracek, great work 👍

aphelionz commented on Dockerfile:
```diff
@@ -0,0 +1,73 @@
+COPY crates/pedersen/Cargo.toml crates/pedersen/Cargo.toml
+COPY crates/pedersen/benches crates/pedersen/benches
+
+RUN RUSTFLAGS='-L/usr/lib -Ctarget-feature=-crt-static' cargo build --target=x86_64-unknown-linux-musl
```
```suggestion
RUN RUSTFLAGS='-L/usr/lib -Ctarget-feature=-crt-static' cargo build --release --target=x86_64-unknown-linux-musl
```

aphelionz commented on Dockerfile:
```diff
@@ -0,0 +1,73 @@
+RUN touch crates/pathfinder/src/lib.rs
+RUN touch crates/pedersen/src/lib.rs
+
+RUN RUSTFLAGS='-L/usr/lib -Ctarget-feature=-crt-static' cargo build --target=x86_64-unknown-linux-musl --bin pathfinder
```
```suggestion
RUN RUSTFLAGS='-L/usr/lib -Ctarget-feature=-crt-static' cargo build --release --target=x86_64-unknown-linux-musl --bin pathfinder
```

aphelionz reviewed (changes requested):
Whoops, actually. My bad. Let's use `--release`

aphelionz commented on Dockerfile:
```diff
@@ -0,0 +1,73 @@
+RUN apk add --no-cache tini
+
+COPY --from=rust-builder ["/usr/lib/libstdc++.so.6", "/usr/lib/libgcc_s.so.1", "/usr/lib/libgmp.so.10", "/usr/lib/" ]
+COPY --from=rust-builder /usr/src/pathfinder/target/x86_64-unknown-linux-musl/debug/pathfinder /usr/local/bin/pathfinder
```
```suggestion
COPY --from=rust-builder /usr/src/pathfinder/target/x86_64-unknown-linux-musl/release/pathfinder /usr/local/bin/pathfinder
```

fracek commented:
I added the `--release` flag.

aphelionz commented:
Great work @fracek we'll look after merging once the tests pass

aphelionz commented:
@Mirko-von-Leipzig the test failure is the issue with GH secrets right? I don't think anything in `.dockerignore` / `README.md` / `Dockerfile` breaks this

CHr15F0x commented:
> the test failure is the issue with GH secrets right?
Yes, it is.

kkovaacs commented on Dockerfile:
```diff
@@ -0,0 +1,73 @@
+COPY --from=python-builder /usr/local/lib/python3.8/ /usr/local/lib/python3.8/
+
+# Create directory and volume for persistent data
+RUN mkdir -p /usr/share/pathfinder/data
```
Hmm, shouldn't this be something like `/var/lib/pathfinder`?

According to the [Linux Filesystem Hierarchy Standard](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch04s11.html):

> The /usr/share hierarchy is for all read-only architecture independent data files. [[30]](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch04s11.html#ftn.idm236091847456)

And for `/var/lib` ([FHS reference](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s08.html)):

> This hierarchy holds state information pertaining to an application or the system. State information is data that programs modify while they run, and that pertains to one specific host. Users must never need to modify files in /var/lib to configure a package's operation, and the specific file hierarchy used to store the data must not be exposed to regular users.

fracek replied:
> This hierarchy holds state information pertaining to an application or the system. State information is data that programs modify while they run, and that pertains to one specific host. **Users must never need to modify files in /var/lib to configure a package's operation**, and the specific file hierarchy used to store the data must not be exposed to regular users.

There should be 2 volumes: one for the configuration in `/usr/share` and one for application data in `/var/lib`.

aphelionz replied:
@fracek do you want to make a PR for those volumes? Also note I just created #190 if you want to give any feedback there

Mirko-von-Leipzig commented:
Thanks @fracek!

kkovaacs reviewed (commented):
LGTM other than the small nit-pick WRT data directory location.
